### PR TITLE
backend/feat: Fleet Management

### PR DESCRIPTION
**Pull request by AshishSaini25**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description

Fleet Management 

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/app/dashboard/Lib/src/API/Dashboard/Registration.hs [hunk 1]
```diff
@@ -40,14 +41,20 @@ type API =
              :> ReqBody '[JSON] DReg.SwitchMerchantReq
              :> Post '[JSON] DReg.LoginRes
        )
+    :<|> "fleet"
+      :> "register"
+      :> ReqBody '[JSON] DReg.FleetRegisterReq
```
Comment (roman-bodavskiy) [inline Backend/app/dashboard/Lib/src/API/Dashboard/Registration.hs:46]: should we use `DashboardAuth` here?

File: Backend/app/dashboard/Lib/src/Tools/Auth/Dashboard.hs [hunk 2]
```diff
@@ -93,3 +93,5 @@ verifyDashboardAccess requiredAccessType personId = do
         else throwError AccessDenied
     DRole.DASHBOARD_USER ->
       pure person.id
+    DRole.FLEET_OWNER ->
```
Comment (roman-bodavskiy) [inline Backend/app/dashboard/Lib/src/Tools/Auth/Dashboard.hs:96]: `DashboardAuth FLEET_OWNER` will success for each dashboard user (the same as `DashboardAuth DASHBOARD_USER`). So, adding new access type `FLEET_OWNER` make no sense, I think

File: Backend/app/dashboard/Lib/src/Domain/Action/Dashboard/Registration.hs [hunk 3]
```diff
@@ -234,3 +256,56 @@ buildRegistrationToken personId merchantId = do
...
+  fleetOwnerRole <- QRole.findByName "FLEET_OWNER" >>= fromMaybeM (RoleDoesNotExist "FLEET_OWNER") --- we have to create a role for the fleet owner
+  fleetOwner <- buildFleetOwner req fleetOwnerRole.id
+  decFleetOwner <- decrypt fleetOwner
+  merchant <-
+    QMerchant.findByShortId req.merchantId
+      >>= fromMaybeM (MerchantDoesNotExist req.merchantId.getShortId)
+  availableServers <- asks (.dataServers)
+  unless (merchant.serverName `elem` (availableServers <&> (.name))) $
+    throwError $ InvalidRequest "Server for this merchant is not available"
+  merchantAccess <- DP.buildMerchantAccess decFleetOwner.id merchant.id
```
Comment (roman-bodavskiy) [inline Backend/app/dashboard/Lib/src/Domain/Action/Dashboard/Registration.hs:279]: no need to decrypt:
` merchantAccess <- DP.buildMerchantAccess fleetOwner.id merchant.id`

File: Backend/app/dashboard/Lib/src/Storage/Tabular/Person.hs [hunk 4]
```diff
@@ -68,8 +70,12 @@ instance ToTType PersonT Domain.Person where
...
       { id = getId id,
         roleId = toKey roleId,
-        emailEncrypted = email & unEncrypted . (.encrypted),
-        emailHash = email.hash,
+        emailEncrypted = case email of
+          Just em -> Just $ em & unEncrypted . (.encrypted)
+          _ -> Nothing,
+        emailHash = case email of
+          Just em -> Just em.hash
+          _ -> Nothing,
```
Comment (roman-bodavskiy) [inline Backend/app/dashboard/Lib/src/Storage/Tabular/Person.hs:78]: emailEncrypted = email <&> (unEncrypted . (.encrypted)),
        emailHash = email <&> (.hash),

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/Fleet.hs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+ the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
+-}
+
+module Domain.Types.Fleet where
+
+import Data.OpenApi (ToSchema)
+import EulerHS.Prelude hiding (id, state)
+
+data FleetOwnerLoginReq = FleetOwnerLoginReq
```
Comment (roman-bodavskiy) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Types/Fleet.hs:20]: These are request types, not domain types. So, they should be placed in `Domain.Action.Dashboard.Fleet.Registration` or even in dashboard  `CommonAPIs` lib`

File: Backend/app/dashboard/Lib/src/Storage/Tabular/Person.hs (hunk 4 above)
Comment (AshishSaini25) [reply Backend/app/dashboard/Lib/src/Storage/Tabular/Person.hs:78]: ok

File: Backend/app/dashboard/Lib/src/Tools/Auth/Dashboard.hs (hunk 2 above)
Comment (AshishSaini25) [reply Backend/app/dashboard/Lib/src/Tools/Auth/Dashboard.hs:96]: ok so here we can use same as done for admin .

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Fleet/Registration.hs [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+import qualified SharedLogic.MessageBuilder as MessageBuilder
+import Storage.CachedQueries.Merchant as QMerchant
+import Tools.Error
+import Tools.SMS as Sms hiding (Success)
+
+---------------------------------------------------------------------
+fleetOwnerLogin ::
+  ( HasFlowEnv m r '["smsCfg" ::: SmsConfig],
+    EsqDBFlow m r,
+    EsqLocDBFlow m r,
```
[the remark this answers is not in the document]
Comment (AshishSaini25) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Fleet/Registration.hs:41]: ok

File: Backend/app/dashboard/Lib/src/Domain/Action/Dashboard/Registration.hs [hunk 7]
```diff
@@ -234,3 +254,55 @@ buildRegistrationToken personId merchantId = do
...
+  ( EsqDBFlow m r,
+    EncFlow m r,
+    HasFlowEnv m r '["dataServers" ::: [Client.DataServer]]
+  ) =>
+  FleetRegisterReq ->
+  m APISuccess
+registerFleetOwner req = do
+  runRequestValidation validateFleetOwner req
+  unlessM (isNothing <$> QP.findByMobileNumber req.mobileNumber req.mobileCountryCode) $ throwError (InvalidRequest "Phone already registered")
+  fleetOwnerRole <- QRole.findByName "FLEET_OWNER" >>= fromMaybeM (RoleDoesNotExist "FLEET_OWNER") --- we have to create a role for the fleet owner
```
Comment (hkmangla) [inline Backend/app/dashboard/Lib/src/Domain/Action/Dashboard/Registration.hs:268]: find by `dashboardAccessType`

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Driver.hs [hunk 8]
```diff
@@ -681,18 +684,40 @@ addVehicle merchantShortId reqDriverId req = do
...
+  -- merchant access checking
+  let merchantId = driver.merchantId
+  unless (merchant.id == merchantId) $ throwError (PersonDoesNotExist driver.id.getId)
+  mbLinkedVehicle <- B.runInReplica $ QVehicle.findByAnyOf (Just req.registrationNo) Nothing
+  mbLinkeddriver <- QVehicle.findById driver.id
+  whenJust mbLinkedVehicle $ \_ -> throwError VehicleAlreadyLinked
+  whenJust mbLinkeddriver $ \vehicle -> throwError $ DriverAlreadyLinkedWithVehicle vehicle.registrationNo
+  let updDriver = driver {DP.firstName = req.driverName} :: DP.Person
+  _ <- QPerson.updatePersonRec driver.id updDriver
+  Redis.set (DomainRC.makeFleetOwnerKey driver.id) fleetOwnerId -- setting this value here , so while creation of creation of vehicle we can add fleet owner id
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Dashboard/Driver.hs:705]: on basis of vehicle number.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Vehicle.hs [hunk 9]
```diff
@@ -56,6 +57,9 @@ upsert a@Vehicle {..} = do
 findById :: (MonadFlow m) => Id Person -> m (Maybe Vehicle)
 findById (Id driverId) = findOneWithKV [Se.Is BeamV.driverId $ Se.Eq driverId]
 
+findByFleetOwnerId :: (MonadFlow m) => Maybe Text -> m [Vehicle]
+findByFleetOwnerId fleetOwnerId = findAllWithKV [Se.Is BeamV.fleetOwnerId $ Se.Eq fleetOwnerId]
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Vehicle.hs:61]: add index in db for this.

File: Backend/dev/migrations/provider-dashboard/0018-fleet-management.sql [hunk 10]
```diff
@@ -0,0 +1,3 @@
+ALTER TABLE atlas_bpp_dashboard.person ALTER COLUMN email_encrypted DROP NOT NULL;
+ALTER TABLE atlas_bpp_dashboard.person ALTER COLUMN email_hash DROP NOT NULL;
+ALTER TABLE atlas_bpp_dashboard.person ALTER COLUMN password_hash  DROP NOT NULL;
```
Comment (hkmangla) [inline Backend/dev/migrations/provider-dashboard/0018-fleet-management.sql:3]: Same for queries for rider dashboard too.

File: Backend/dev/migrations/dynamic-offer-driver-app/0263-add-fleet-owner-in-vehicle.sql [hunk 11]
```diff
@@ -0,0 +1 @@
+ALTER TABLE atlas_driver_offer_bpp.vehicle ADD COLUMN fleet_owner_id TEXT;
```
Comment (hkmangla) [inline Backend/dev/migrations/dynamic-offer-driver-app/0263-add-fleet-owner-in-vehicle.sql:1]: varying character(36)

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/dashboard/CommonAPIs/src/Dashboard/ProviderPlatform/Driver.hs [hunk 12]
```diff
@@ -662,6 +662,41 @@ instance HideSecrets AddVehicleReq where
...
+  "getAllVehicle"
+    :> "fleet"
+    :> Get '[JSON] ListVehicleRes
+
+newtype ListVehicleRes = ListVehicleRes
+  {vehicles :: [VehicleAPIEntity]}
+  deriving (Generic, ToJSON, ToSchema, FromJSON)
+
+data VehicleAPIEntity = VehicleAPIEntity
+  { driverId :: Text,
```
Comment (max-keviv) [inline Backend/app/dashboard/CommonAPIs/src/Dashboard/ProviderPlatform/Driver.hs:689]: @AshishSaini25 is there a seperate api to get driver status, earning , trips completed

**Pull request merged**